Add source support to regex and json stages

pracucci opened the pull request:
**What this PR does / why we need it**:

Following up the idea shared in https://github.com/grafana/loki/issues/703, in this PR I'm proposing to add a config property called `source` both to `regex` and `json` stages. When `source` is valued, the regex/json parsing is done based on `extracted[source]` instead of the default `entry`.

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/loki/issues/703

**Checklist**
- [x] Documentation added
- [x] Tests updated

slim-bean commented on pkg/logentry/stages/json.go:
```diff
@@ -24,6 +25,7 @@ const (
 // JSONConfig represents a JSON Stage configuration
 type JSONConfig struct {
 	Expressions map[string]string `mapstructure:"expressions"`
+	Source      string            `mapstructure:"source"`
```
For consistency, we've been handling optional config strings as a pointer to string

slim-bean commented on pkg/logentry/stages/json_test.go:
```diff
@@ -31,28 +42,51 @@ var testJSONLogLine = `
-	if err != nil {
-		t.Fatal(err)
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
```
Can you set this up for this test to run in parallel?  Like the `TestJSONParser_Parse`

slim-bean commented on pkg/logentry/stages/regex.go:
```diff
@@ -21,6 +22,7 @@ const (
 // RegexConfig contains a regexStage configuration
 type RegexConfig struct {
 	Expression string `mapstructure:"expression"`
+	Source     string `mapstructure:"source"`
```
samer as JSON, should be a pointer to string

slim-bean commented on pkg/logentry/stages/regex_test.go:
```diff
@@ -13,42 +13,84 @@ import (
-	if err != nil {
-		t.Fatal(err)
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
```
also would be nice for this guy to be parallel too

slim-bean commented:
This look really good @pracucci, a couple nits but otherwise looks great!

pracucci commented:
Thanks @slim-bean for your review! I should have addressed your feedback:

- [Converted source in regex and josn stages to string pointer](https://github.com/grafana/loki/pull/735/commits/551221057b7dd7c41f0166f832a0e2a35bb06441)
- [Added source validation to regex and json stages](https://github.com/grafana/loki/pull/735/commits/7af95e0acb58832f5ba76c3a1ec5c9395045d502)
- [Parallelly run regex and json stages pipeline tests](https://github.com/grafana/loki/pull/735/commits/428d5a24eefb685327091d2b786dc38725bdd6fc)

slim-bean commented:
LGTM! Thanks @pracucci
